tlair/src/main/scala/at/forsyte/apalache/tla/lir/TypeLevelBuilder.scala

```diff
@@ -0,0 +1,186 @@
+package at.forsyte.apalache.tla.lir
+
+import at.forsyte.apalache.tla.lir.oper.{TlaArithOper, TlaBoolOper, TlaFunOper, TlaOper}
+
+// We use shapeless for facilitating the type-level encodings of TlaExpressions
+import shapeless._
+import poly._
+
+// A (very) rough sketch at what TLA Expressions might look like when
+// we represent out typing discipline inside of Scala's type system.
+object Tla {
+  // The type of any TLA expression
+  sealed trait Exp[T] {
+    def toString: String
+  }
+
+  // Just an alias
+  type IntT = BigInt
+
+  // A value constructs a TLA expression tagged with the Scala type it carries as a payload.
+  class Value[T](value: T) extends Exp[T]
+  case class Int(value: IntT) extends Value(value)
+  case class Str(value: String) extends Value(value)
+  case class Bool(value: Boolean) extends Value(value)
+
+  // Each compound expression is reprsented by a new type
+
+  sealed trait SeqType[T]
+  case class TlaSeq[T](seq: Seq[Exp[T]]) extends Exp[SeqType[T]]
+
+  sealed trait TupleType[A, B]
+  case class Empty() extends Exp[TupleType[Unit, Unit]]
+  case class Single[A](fst: A) extends Exp[TupleType[A, Unit]]
+  case class Tuple[A, B](fst: A, snd:B) extends Exp[TupleType[A, B]]
+
+  sealed trait RecType[Fields]
+  case class Rec[Fields <: HList](fields: Fields) extends Exp[RecType[Fields]]
+
+  sealed trait OperType[Args <: HList, Ret]
+
+  // Names are expressions
+  // The idea here is that names also record the type of the expression
+  // they refer to. When we don't know what that type is, we might use
+  // `Name[Exp[Any]]`.
+  case class Name[T](name: String) extends Exp[T]
```

Comment:
Is there a way to refer to an identifier like `:foo` in lisp?

Reply:
I'm not sure exactly how that would factor in. Perhaps give an example?

Reply:
I am not sure either. Actually, TLA+ identifiers may have the form `Foo!Bar`, when they are instantiated from modules. So maybe this does not matter. It would probably matter when we have to describe the shape of records.

Reply:
But I see that you are using `Fields` for the records. I am not sure how it works.

Reply:
I'm not clear what the result you're wanting from identifiers/fields here. What would you like to do in relation to to `Name`s?